neuraxle/pipeline.py

```diff
@@ -62,79 +34,158 @@ def __hash__(self):
 
-    def fit_transform(self, data_inputs, expected_outputs=None) -> ('ResumablePipelineRunner', Any):
-        for step_name, step in self.steps_as_tuple[:-1]:
-            # TODO: overwrite step's self return value on fitting.
-            data_inputs = step.fit_transform(data_inputs, expected_outputs)
-        processed_outputs = self.steps_as_tuple[-1][-1].fit_transform(data_inputs, expected_outputs)
-        return processed_outputs
+    def __init__(self, steps: NamedTupleList):
+        BaseStep.__init__(self)
+        TruncableSteps.__init__(self, steps)
 
-    def fit(self, data_inputs, expected_outputs=None) -> 'ResumablePipelineRunner':
-        for step_name, step in self.steps_as_tuple[:-1]:
-            # TODO: overwrite step's self return value on fitting.
-            data_inputs = step.fit_transform(data_inputs, expected_outputs)
-        processed_outputs = self.steps_as_tuple[-1][-1].fit(data_inputs, expected_outputs)
-        return processed_outputs
+    @abstractmethod
+    def fit(self, data_inputs, expected_outputs=None) -> 'BasePipeline':
+        raise NotImplementedError()
 
+    @abstractmethod
     def transform(self, data_inputs):
+        raise NotImplementedError()
 
-        dos = [DataObject(i, di) for i, di in enumerate(data_inputs)]
+    @abstractmethod
+    def fit_transform(self, data_inputs, expected_outputs=None) -> ('BasePipeline', Any):
+        raise NotImplementedError()
 
-        # TODO: review this and do it in fit also.
-        steps = self.resume_transform_entry_point(dos, self.steps_as_tuple)
+    @abstractmethod
+    def inverse_transform_processed_outputs(self, data_inputs) -> Any:
+        raise NotImplementedError()
 
-        for step_name, step in steps:
-            data_inputs = step.transform(data_inputs)
-        processed_outputs = data_inputs
-        return processed_outputs
+    def inverse_transform(self, processed_outputs):
+        if self.transform != self.inverse_transform:
+            raise BrokenPipeError("Don't call inverse_transform on a pipeline before having mutated it inversely or "
+                                  "before having called the `.reverse()` or `reversed(.)` on it.")
 
-    def resume_transform_entry_point(self, dos, steps_as_tuple):
-        steps = copy(steps_as_tuple)
-        for step_name, step in reversed(steps_as_tuple):
+        return self.inverse_transform_processed_outputs(processed_outputs)
 
-            # TODO: maybe change to: if hasattr(step, 'should_resume'):
-            if isinstance(step, BaseBarrier) or isinstance(step, Pipeline):
 
-                if step.should_resume(dos):
-                    if isinstance(step, Pipeline):  # TODO: maybe change to: if hasattr(step, 'trim_for_resume'):
-                        step = step.trim_for_resume(dos)
-                        steps[step_name] = step
-                    steps = steps_as_tuple[step_name:]  # TODO: fix get
-                    return steps
+class ResumableStep(BaseStep):
```

Comment:
I think this is a mixin, not inheriting from BaseStep. Diamond OOP problem seems to happen when declaring `class Pipeline(..., ...)`